### PR TITLE
Native implementation of Shrake-Rupley algorithm for atomic solvent accessible surface areas.

### PR DESCRIPTION
**Pull request by JoaoRodrigues**
- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

This PR implements the standard Shrake-Rupley algorithm to calculate atomic solvent accessible surface areas (SASAs/ASAs) using the KDTree data structure. This allows Biopythoneers to calculate this metric without a third-party application (e.g. NACCESS). The use of KDTrees makes this calculate quite competitive with other approaches out there that use Cython/C extensions.

TODO:
- [x] Working implementation
- [x] Test/Improve API.
- [x] Extend vdW radii list.
- [x] Documentation
- [x] Examples
- [x] Tests & Unit Tests

I'd appreciate if someone can have a look at the code and see if there are any glaring issues or possible improvements to both speed and readability. Pinging @mdehoon (would it make sense to change the bucket size of the KDTree here?) & @jgreener64 for insights.

~~Also, @peterjc , from an API point of view, I'd rather assign an `atom.asa` attribute to each atom but other modules use the `atom.xtra` dictionary. Any preferences here? I'll be adding some more methods to for instance, aggregate the values at residue/chain level, so nothing is set yet.~~ *EDIT:* opted for the attribute, seemed the clearest during testing.

### CODE REVIEW

**Comment by mdehoon** [comment]
@JoaoRodrigues  Actually I don't know much about the algorithm behind KDTrees. I modified the code at some point for the Python2 -> Python3 transition and to clean up the C code, but that is about it. I will rely on your judgement here.

**Comment by JoaoRodrigues** [comment]
I've benchmarked this against freesasa and it gives the same values on several structures. Just pushed a merge commit to update to the latest master and then will merge.

**Comment by jgreener64** [comment]
Looks cool, I don't have time to look at this in detail at the moment though.

**Comment by JoaoRodrigues** [comment]
👍 No problem @jgreener64, thanks!

File: Bio/PDB/__init__.py [hunk 1]
```diff
@@ -72,3 +81,10 @@
     from .NeighborSearch import NeighborSearch
 except ImportError:
     pass
+
+# Native Shrake-Rupley algorithm for SASA calculations.
+# Depends on kdtrees C module
+try:
+    from .SASA import ShrakeRupley
```
Comment (peterjc) [inline Bio/PDB/__init__.py:88]: Why import this here? Admittedly Bio.PDB has a lot of historic automatic import, but both the tests and the NEWS entry suggest you'd use:

```python
from Bio.PDB.SASA import ShrakeRupley
```

File: Bio/PDB/__init__.py [hunk 2]
```diff
@@ -10,7 +10,16 @@
...
+Original Author: Thomas Hamelryck.
+Contributions by:
+- Peter Cock
+- Joe Greener
+- Rob Miller
+- Lenna X. Peterson
+- Joao Rodrigues
+- Kristian Rother
+- Eric Talevich
+- and many others.
```
Comment (peterjc) [inline Bio/PDB/__init__.py:22]: I think having module contributors other than the original author(s) listed in the docstring is unusual for Biopython (usually author and major contributors' names are in the copyright block at the top of each file). Although that doesn't mean we can't change this...

**Review by peterjc — COMMENTED** [review]
Sorry, didn't realise my lack of review was holding this up. A comment on contributor names in docstrings, and a strong suggestion on removing the automatic import.

**Comment by peterjc** [comment]
Ah, already merged - that explains why the GitHub UI seemed a little off.

Still, do you think it is worth removing that import?

File: Bio/PDB/__init__.py (hunk 1 above)
Comment (JoaoRodrigues) [reply Bio/PDB/__init__.py:88]: For simplicity, I can remove it of course!

**Comment by JoaoRodrigues** [comment]
Sorry, I waited a few days before merging and since it all seemed fine from a functionality point of view ... 

Yeah, I can remove the top-level import, no issue!

**Comment by peterjc** [comment]
You had Michiel and Joe look at it, merging it was fine. I only spotted this later.

**Pull request merged**